### PR TITLE
Cache LLVM binaries in repo instead of downloading (xplat stage 1)

### PR DESCRIPTION
**Pull request by swernli**
This change puts the Win32 LLVM binaries into the repo and removes the infrastructure for downloading them from an external source. This also includes some clean-up of unused cmdlets from the CommonBuild module.

These changes are the first stage of introducing the cross-platform build (as requested in #193) and originally part of the large PR at #205

### CODE REVIEW

File: PsModules/CommonBuild/CommonBuild.psm1 [hunk 1]
```diff
@@ -49,16 +49,6 @@ function Get-DefaultBuildPaths([string]$repoRoot)
     return $buildPaths
 }
 
-function Update-Submodules
```
Comment (smaillet) [inline PsModules/CommonBuild/CommonBuild.psm1:52]: Please don't remove these, it's called CommonBuild for a reason, it's used on multiple repos, not every repo will use all functions but that doesn't mean they have no value.

File: PsModules/CommonBuild/CommonBuild.psm1 [hunk 2]
```diff
@@ -122,67 +112,6 @@ function ConvertTo-PropertyList([hashtable]$hashTable)
     return ( $hashTable.GetEnumerator() | %{ @{$true=$_.Key;$false= $_.Key + "=" + $_.Value }[[string]::IsNullOrEmpty($_.Value) ] } ) -join ';'
 }
 
-function Invoke-TimedBlock([string]$activity, [ScriptBlock]$block)
```
Comment (smaillet) [inline PsModules/CommonBuild/CommonBuild.psm1:125]: Restore this please.

File: PsModules/CommonBuild/CommonBuild.psm1 [hunk 3]
```diff
@@ -122,67 +112,6 @@ function ConvertTo-PropertyList([hashtable]$hashTable)
...
-    Input stream containing compressed ZIP archive data to expand
-
-.PARAMETER OutputPath
-    Out put destination for the decompressed data
-#>
-    $zipArchive = [System.IO.Compression.ZipArchive]::new($src)
-    [System.IO.Compression.ZipFileExtensions]::ExtractToDirectory( $zipArchive, $OutputPath)
-}
-
-function Expand-StreamFromUri([Parameter(Mandatory=$true, ValueFromPipeLine)]$uri, [Parameter(Mandatory=$true)]$OutputPath)
```
Comment (smaillet) [inline PsModules/CommonBuild/CommonBuild.psm1:170]: Restore this

File: PsModules/CommonBuild/CommonBuild.psm1 [hunk 4]
```diff
@@ -324,60 +253,6 @@ function Invoke-MSBuild([string]$project, [hashtable]$properties, $targets, $log
     }
 }
 
-function Find-7Zip()
```
Comment (smaillet) [inline PsModules/CommonBuild/CommonBuild.psm1:327]: restore this

File: llvm/Llvm-Libs.props [hunk 5]
```diff
@@ -0,0 +1,179 @@
+<Project>
```
Comment (smaillet) [inline llvm/Llvm-Libs.props:1]: Why was this file added, what does it do, how does it play in the new design? At the very least this should contain comments explaining the role this plays, and a readme.md with all the gory details.

File: Move-LlvmBuild.ps1 [hunk 6]
```diff
@@ -0,0 +1,18 @@
+$plat = "win32"
+$basePath = Join-Path $PSScriptRoot llvm
+
+# Copy from cached headers.
+Write-Information "Unzipping include files..."
+Expand-Archive -LiteralPath (Join-path $basePath xplat $plat include.zip) -DestinationPath $basePath -Force
+
+# Copy from cached binaries.
+$libDir = Join-Path $basePath lib
```
Comment (smaillet) [inline Move-LlvmBuild.ps1:9]: Braces on their own line please.

File: Move-LlvmBuild.ps1 [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+if (!(Test-Path $libDir)) {
```
Comment (smaillet) [inline Move-LlvmBuild.ps1:10]: Braces on their own line

File: Move-LlvmBuild.ps1 [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    New-Item -ItemType Directory $libDir
+}
```
Comment (smaillet) [inline Move-LlvmBuild.ps1:12]: Braces on their own line

File: Move-LlvmBuild.ps1 [hunk 9]
```diff
@@ -0,0 +1,18 @@
+$plat = "win32"
```
Comment (smaillet) [inline Move-LlvmBuild.ps1:1]: This isn't so much "moving" as it is unpacking, so let's call it Expand-LlvmBuild for consistency with standard PowerShell verbs

File: Build-Interop.ps1 [hunk 10]
```diff
@@ -76,13 +79,7 @@ try
     }
     #</HACK>
 
-    # Download and unpack the LLVM libs if not already present, this doesn't use NuGet as the NuGet compression
```
Comment (smaillet) [inline Build-Interop.ps1:79]: Please keep the bulk of this comment as it still applies, only difference is that it doesn't download from a separate location.

File: .gitignore [hunk 11]
```diff
@@ -155,7 +155,10 @@ Tools/
 *.binlog
 *.7z
 
-/llvm
```
Comment (smaillet) [inline .gitignore:158]: We need to have a git attributes file as well to set files as text by default, but the .zip files are bin files.

There is already a problem where GitHub thinks the PSD1 in the PR is a Binary file and not text, so that needs to be corrected as well. (I can't diff that file or force it from the GH review page, and It sadly won't even allow me to make a comment on the file.) 

Update: I've created #211 to track the underlying issue with that file and the RC file as they both have a UTF-16LE BOM, which git treats as binary.

Having said that and had time to review and think about this all, I'm concerned by the checking in of binaries to the repo. Not only does it cause issue with respect to correct attribution of a binary for the tooling (and incorrectly processing line ending in a binary), it makes it difficult to properly access control those files. (E.g. no-one should ever checkin these files, that should only be done by automated tooling so their authenticity is guaranteed.) That's why the original plan for this cross-plat stuff was to actually build the libs from source. Since that turns out to be impractical. I think it is best to stick with the previous approach using a separate repo for the LLVM libs and includes.

File: Move-LlvmBuild.ps1 [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+# Copy from cached headers.
+Write-Information "Unzipping include files..."
+Expand-Archive -LiteralPath (Join-path $basePath xplat $plat include.zip) -DestinationPath $basePath -Force
+
+# Copy from cached binaries.
+$libDir = Join-Path $basePath lib
+if (!(Test-Path $libDir)) {
+    New-Item -ItemType Directory $libDir
+}
+foreach ($zip in (Get-ChildItem (Join-Path $basePath xplat $plat) -Recurse -Include "lib.*.zip")) {
```
Comment (smaillet) [inline Move-LlvmBuild.ps1:13]: Why are these named Lib1, Lib2, etc...? That seems rather generic and non-descript. Are they a multi-file zip that requires such a naming strategy? Can we at least use a prefix, that specifies the details that these are LLVM libs, etc.. and the version? Ideally these would break down to the compilation tool and target arch. So that someone only interested in building for a particular target or tool wouldn't need everything.

**Review by smaillet — CHANGES_REQUESTED** [review]
First off, thanks for all the hard work in figuring out how to get to crossplat build of this project, it is something I've wanted for a while but lacked the expertise and resources to figure out.

There are a few things to work out on this, the obvious elephant in the room is the LLVM lib binaries and build. When we first discussed this effort the idea was to leverage additional backend support to build the LLVM libraries on every build. Unfortunately, that is still a very time consuming process and worse, wasteful on every build as the LLVM code changes so rarely (about twice a year with current cadence from LLVM). So, this change, doesn't do the build. Instead it checks in the large zip files as binaries. After some consideration and review I see there are issues with this approach:

1. How does one even review a binary ZIP file?
2. How do we protect the binaries to ensure that ONLY officially built binaries are included in the repo?
3. Can we just place the LLVM xplat build into the existing LLVM libs build repo and tooling to publish the binary artifacts as was done previously? (Since the plan of always building the libs didn't pan out, I'm not seeing an advantage to checking in the binaries, and I see several drawbacks)

Additionally
* There are a lot of things removed from CommonBuild module that have no reason to be removed. Please restore them all. The goal is that PS module will grow and integrate functionality across multiple repos parameterizing repo specific behavior. Additionally, the plan is to make it so it is available via PSGet instead of a copy in every repo so it is truly common. 
* The Stages seem backwards in ordering. Stage 1 includes binaries of the zips without any means to actually build them etc... The ask on breaking down the large 90+ file PR was to focus each PR on a specific change of functionality. For example, change the build support of the native code to use CMAKE as one PR (even if it only targets Windows at first), another to update the native library load/pinvoke support to allow xplat loading (again even if it only builds for Windows at first), etc... These sorts of targeted structural changes have value on their own and don't require all the xplat support, but set the stage for the ultimate xplat goal. Keeping PRs small and focused allows for better evaluation of the actual changes, allows for easier reverts if something goes wrong, and allows for acceptance of individual PRs while debating/discussing others. (There's lots of good stuff in these PRs, but it's pretty much an all or nothing call on them and as stated above I've reconsidered whether binary check-ins are the way to go.)

File: Move-LlvmBuild.ps1 (hunk 12 above)
Comment (smaillet) [reply Move-LlvmBuild.ps1:13]: Where did these binary zip files come from and how do I validate they are legit? (This is a fundamental issue of executables/ZIP format binaries in a source control system)

**Review by smaillet — CHANGES_REQUESTED** [review]
Generally speaking, if we go the route of checked in ZIP files, this should be one of the later changes, as opposed to the first. (Though I'm no longer a fan of the idea of the checked in binaries as a solution to the LLVM "monster build" problem so the path forward there isn't clear)

The changes to build LibLLVM via CMAKE, and the changes to the library loading are two sets of changes that are independent of one another and have value on their own.

**Comment by swernli** [comment]
@smaillet Thanks for taking the time to review! Given your concerns about the checked in binary approach, I'm going to explore an alternative solution.

**Comment by smaillet** [comment]
OK, I', interested in any alternatives you can think of. The problem of the LLVM libs has been a giant thorn of this project from the start. So far I can only think of a few core options:

1. Build everything always
   1. This is, of course the ideal, but rather wasteful as the build of LLVM libraries is expensive. 
2. Build the libraries in an isolated pass caching the results "somewhere"
   1. There are a few variations on this theme, generally differentiating on the isolated pass and the location of the cached results
      1. isolation
         1. Build from a distinct repo, uncluttered by anything other than the build of the libraries.
            1. This is obviously how things have worked for a bit now with this library, it's far from ideal but is known to be workable,  the process is understood, and hasn't proven overly painful.
            2. Having an automated build process that could build and publish from here would be a HUGE step in eliminating much of the pain of this approach. So I think that shifting to this warrants some thought.
         2. Build from within the same repo
            1. Same branch
               1. Advantage with this is that the LLVM headers can be referenced in place and only the actual built binaries need be cached. Though the headers don't account for much in the final ZIP files.
            2. unique detached branch, like gh-pages
               1. Doesn't seem to have any advantage over another repo and actually is more complicated to deal with, in general.   
      2. Caching of output
         1. This is where things have room for exploration... The current approach of publishing the results as a release on GitHub, works but is painful. An alternative I considered early on is to leverage the build system to cache the data for us. Many do have a way to specify "artifacts" that can be retained for use in future builds. Thus, if we had a way for the build/cloud infrastructure to cache specific output of the build then the build itself could hold everything internally. We'd only need to update the encapsulated binaries a few times a year so wouldn't need the full LLVM build all the time.  I had discarded this approach due to limits of public hosting services like AppVeyor the repo leveraged at the time (It has such a facility but is limited to a size below what is needed by the LLVM libraries)
